src/recovery/tally.rs

```diff
@@ -267,29 +279,51 @@ impl<I: Iterator<Item = PersistentFragmentLog>> VoteFragmentFilter<I> {
             timeframe,
             era,
             fragments,
+            fees,
             replay_protection: HashSet::new(),
             spending_counters: HashMap::new(),
         })
     }
 
-    fn validate_tx<P>(
+    fn validate_tx<P: chain_impl_mockchain::transaction::Payload>(
         &mut self,
-        identifier: &account::Identifier,
-        witness: &account::Witness,
         transaction: &TransactionSlice<P>,
-    ) -> (bool, SpendingCounter) {
+        fragment_id: FragmentId,
+    ) -> Result<SpendingCounter, ValidationError> {
+        let (_, identifier, witness) = deconstruct_account_transaction(&transaction)?;
+
+        transaction.verify_strictly_balanced(self.fees.calculate_tx(transaction))?;
+
         let spending_counter = self
             .spending_counters
             .entry(identifier.clone())
             .or_default();
-        verify_original_tx(
+
+        let (valid, sc) = verify_original_tx(
             SpendingCounter::from(*spending_counter),
             &self.block0.into_hash(),
             &transaction.transaction_sign_data_hash(),
-            identifier,
-            witness,
+            &identifier,
+            &witness,
             self.range_check.clone(),
-        )
+        );
+
+        if !valid {
+            return Err(ValidationError::InvalidTransactionSignature {
+                id: fragment_id.to_string(),
+                range: self.range_check.clone(),
+            });
+        }
+
+        // check if fragment was processed already
```

Comment:
I think it would make sense to check this before doing all other checks. Even before deconstructing a transaction.